Add building-a-pure-silent-payment-mobile-wallet-challenges-and-hindsight review to bitcoinplusplus/sovereignty-edition (#701)
bitcointranscripts/bitcointranscripts: bitcoinplusplus/sovereignty-edition/building-a-pure-silent-payment-mobile-wallet-challenges-and-hindsight.md

---
title: 'Building a Sovereign Silent Payments Mobile Wallet: Making Silent Payments Practical'
transcript_by: 'muchai254 via review.btctranscripts.com'
media: 'https://youtu.be/MgoPkXHUH0E'
date: '2025-12-15'
tags:
  - 'silent-payments'
  - 'wallet'
  - 'privacy'
  - 'cryptography'
  - 'compact-block-filters'
  - 'btcplusplus'
speakers:
  - 'Cygnet'
categories:
  - 'conference'
source_file: 'https://youtu.be/MgoPkXHUH0E'

> summary: 'Cygnet presents Dana Wallet, a pure silent payment mobile wallet built with co-developer Sofstein, covering BIP-352 fundamentals (ECDH-derived unique addresses via input-hash-summed-private-keys times recipient scan key), the two core receiver-side challenges for light clients (needing per-transaction tweak data and scanning every taproot output in every block), and two implementation iterations — an initial Electrum-Rust plus Nakamoto BIP-158 approach and a more mobile-friendly Blindbit Oracle over HTTP. Benchmark results show a budget Motorola phone scanning 16 months of chain history in 6 minutes, made feasible by the finding that 95% of UTXOs are spent within two months so old fully-spent transactions can be skipped, with the talk closing on BIP-353 human-readable addresses as a near-term UX improvement that could eliminate displaying raw Bitcoin addresses entirely.'

Reviewer: ```suggestion
summary: 'Cygnet presents Dana Wallet, a pure silent payment mobile wallet built with co-developer Sosthene, covering BIP-352 fundamentals (ECDH-derived unique addresses via input-hash-summed-private-keys times recipient scan key), the two core receiver-side challenges for light clients (needing per-transaction tweak data and scanning every taproot output in every block), and two implementation iterations — an initial Electrum-Rust plus Nakamoto BIP-158 approach and a more mobile-friendly Blindbit Oracle over HTTP. Benchmark results show a budget Motorola phone scanning 16 months of chain history in 6 minutes, made feasible by the finding that 95% of UTXOs are spent within two months so old fully-spent transactions can be skipped, with the talk closing on BIP-353 human-readable addresses as a near-term UX improvement that could eliminate displaying raw Bitcoin addresses entirely.'
```

Reviewer: @0tuedon Is "summary" generated by AI?  How to change the summary in editor UI? Can we change it directly from the file? The correct dev name is "Sosthene"

https://github.com/Sosthene00
https://github.com/cygnet3/dana

Reviewer 2: No we can't currently change it in this current UI, but  can add a final commit to change this or you can when other checks passes

Reviewer: All green and ready to merge on my end! @0tuedon , any thoughts or feedback?

So, yeah, this is just a big problem, and this is something that silent payments is trying to tackle, basically.

## How Silent Payments Use ECDH

Cygnet: 00:03:04

So the way that if I wanted to describe silent payments in a single sentence, I would say that silent payments is trying to counter address reuse with the use of ECDH.
So ECDH, elliptic curve Diffie-Hellman, it is a cryptographic scheme that is pretty popular, and it is like a scheme that's two parties that have a public-private key, can sort of without really even interacting with each other, they can form a shared secret together.
And this requires private and public key pairs, and as it happens, most Bitcoin transactions tend to use public-private keys.
So if you think of a normal Bitcoin, a common Bitcoin script is like sending to Taproot or sending to like a Witness Public Key Hash (P2WPKH), you're basically sending to a public key, and when you spend it, you are actually using the private key to spend it.
So there's private-public keys here.

> And so the idea of silent payments is that we can exploit this fact, the fact that a lot of transactions on chain are using private-public keys to create a new address.

Reviewer: ```suggestion
And so the idea of silent payments is that we can exploit this fact, the fact that a lot of transactions on-chain are using private-public keys to create a new address.
```

So this function you can see here, that is kind of in a nutshell what silent payments is doing.

> So normally you would kind of send to a public key, which is I call P_recipient, but now instead of sending it to that directly, what I'm going to do is I'm going to use Diffie-Hellman to calculate the shared secret, and I'm not going to send to the address directly, but I'm going to tweak the address with the shared secret, and I'm then going to send it to that address.

Reviewer: ```suggestion
So normally you would kind of send to a public key, which is I call `P_recipient`, but now instead of sending it to that directly, what I'm going to do is I'm going to use Diffie-Hellman to calculate the shared secret, and I'm not going to send to the address directly, but I'm going to tweak the address with the shared secret, and I'm then going to send it to that address.
```

There's a bunch of other reasons for it, too.
But that's short of it.

## Sending and Receiving Silent Payments

Cygnet: 00:05:21

So yeah, that was silent payments in a nutshell.
I'll just quickly go over how you can actually do a silent payment.
So first of all, if you're a sender, you need to do use or spend from an output that is using public keys, or using public-private key pairs
So these are some script types which, again, this is not a very high requirement because these are very popular script types.
So you need to have at least one of these in your spending transaction.
Then using the private keys from the inputs that you're trying to spend, you sum up the private keys.
The reason you sum them up and not just take one at random is because this is slightly more convenient or this is more convenient to do, like, collaborative transactions.

> So you sum up, you create this a, which is the sum of the private keys.

Reviewer: ```suggestion
So you sum up, you create this `a`, which is the sum of the private keys.
```

Next you also need an input hash.
So the reason here that we take an input hash is because just using the sum of the private keys, it's possible to create transactions reusing the same keys.

> So like you can imagine if you have if you already have like an address and you spend from that address twice, you can you're kind of reusing the keys, so just taking the creating a secret just from the private keys itself doesn't guarantee uniqueness, but we guarantee this by taking this input hash which is essentially just the outpoints and then hash, and because outpoints are related to UTXOs, and UTXOs can, of course, only be spent once.

Reviewer: ```suggestion
So like you can imagine if you have if you already have like an address and you spend from that address twice, you can you're kind of reusing the keys, so just taking the creating a secret just from the private keys itself doesn't guarantee uniqueness, but we guarantee this by taking this input hash which is essentially just the outpoints and then hashed, and because outpoints are related to UTXOs, and UTXOs can, of course, only be spent once.
```

Reviewer: <img width="1911" height="1001" alt="Image" src="https://github.com/user-attachments/assets/4f6c8807-947d-4438-92df-c2cfe16b192d" />

If UTXOs could be spent multiple times, then that would be like a double spending problem.
So that is essentially that kind of guarantees that this address that we create is unique.
So next we create a shared secret.
This is the input hash times the A sum times the B_scan, which is the scan key for the recipient that you're trying to send to.
And then finally, we calculate what the resulting on chain output looks like with this formula.

> So P is you have the shared secret, so you send to the B_spend, which in the previous slides would be what you would normally send to, but now we don't just send to the B_spend, we also send to a B_spend summed up with the hash of the shared secret multiplied by the generator points and make it a public key.

Reviewer: ```suggestion
This is the input hash times the `a` sum times the `B_scan`, which is the scan key for the recipient that you're trying to send to.
And then finally, we calculate what the resulting on-chain output looks like with this formula.
So `P` is you have the shared secret, so you send to the `B_spend`, which in the previous slides would be what you would normally send to, but now we don't just send to the `B_spend`, we also send to a `B_spend` summed up with the hash of the shared secret multiplied by the generator point and make it a public key.
```

Reviewer: Should "a" be lowercase?

> This is the input hash times the `a` sum times 

<img width="2395" height="1481" alt="Image" src="https://github.com/user-attachments/assets/ee54d979-bf9d-4c3d-8184-8b0799aa7d53" />

> **`A` must be lowercase `a`:**
   Because this paragraph describes the **sender's** actions ("the recipient *you're trying to send to*"), the sender is summing up their own input **private keys**. Private keys are scalars and must be represented by the lowercase **`a`**. The capital **`A`** represents the sum of the public keys, which is only used on the *receiving* side.

Reviewer: >  **"generator points" to "generator point":**
>    In elliptic curve cryptography (specifically Secp256k1 used in Bitcoin), there is only a single, standardized base point called the **generator point** (usually denoted as `G`). Changing this to the singular "generator point" is technically accurate.

Author: Yeah, it should. Even in the presentation it was lowercased. Lemme correct this

So that is what makes it unique.

> Then on the receiving side, this is this looks actually very similar to the sending side, that is because, you know, it is actually following the same steps, but now instead of taking you're sort of doing it from the perspective of the receiver, so on the input side you're actually looking at the public keys which is shown with a capital A, so capital A is the sum of the input public keys, and now we have this b_scan which is like a small b, so that's the private key for the scan key.

Reviewer: ```suggestion
Then on the receiving side, this is this looks actually very similar to the sending side, that is because, you know, it is actually following the same steps, but now instead of taking you're sort of doing it from the perspective of the receiver, so on the input side you're actually looking at the public keys which is shown with a capital `A`, so capital `A` is the sum of the input public keys, and now we have this `b_scan` which is like a small `b`, so that's the private key for the scan key.
```

So however, the A is the same, the input hash is the same because it follows the same principle, and the shared secret is also essentially the same because this is the Diffie-Hellman scheme.

> That also means that in step five, the resulting on chain is calculated, you basically arrive at the same result as the sender would.

Reviewer: ```suggestion
That also means that in step five, the resulting on-chain is calculated, you basically arrive at the same result as the sender would.
```

Of course, the right side, the two parties know the shared secret, but only the receiver knows the private key, knows the spending key, so only the receiver is actually able to spend this output.
So this is basically how you can do both sending and receiving.

## The Main Challenge: Efficient Wallet Scanning

Cygnet: 00:09:23

In a nutshell, or the way I've just described, it might sound to you like, okay, this seems pretty straightforward, so what is the catch here?
Because if it was that easy to solve address reuse, then why haven't we done this like 10 years ago?
So the catch here is that this process, especially for the receiver side, is very incompatible, or it looks like it's very incompatible with light clients.
So on the sending side, it is relatively straightforward.
All of the extra data that you need is, it's already part of the transaction.
You need to do this extra, like, calculate the shared secret, but other than that, it's basically not that much extra unique stuff.
However, on the receiving side, there are two problems that you need to solve before you can, you know, find these addresses.

> So one of them is that you need extra on chain data.

Reviewer: ```suggestion
So one of them is that you need extra on-chain data.
```

So we have both this input hash, and we have the sum of public keys.
This is part of a transaction.
Normally, light clients, they basically just have a list of addresses that they're interested in, and they have their own ways of finding, like, if there are any funds on those addresses.
So maybe they query an Electrum server, or maybe they use a compact block filter or something.
But the point is they kind of know what scripts they're looking for.
However, in our case, we don't know what those scripts look like.
We have to calculate them.

> And we need on chain data for this, or we need this extra piece of data.

Reviewer: ```suggestion
And we need on-chain data for this, or we need this extra piece of data.
```

You just have to brute force basically every single transaction there is.
So this creates a huge extra what we call a scanning step.
And we need to, like, you know, you can kind of imagine when silent payments was first proposed, that this was considered a death sentence, that this seems like it's way too much.

## Making Silent Payments Work for Light Clients

Cygnet: 00:12:40

However, I think for both of these problems, there are some mitigations.
So the first one is actually not that big of a deal.
We need some extra bit of on chain data.
We need two things.
We need the input hash, and we need the public key, or the sum of the public keys.
So this is something that is essentially all public information, which means that if you want, you can have like some sort of a server that can, for a given block or whatever, can create like an end point that gives this data to you.
So this is called a tweak data, or a tweak.
This is something that is unique for every transaction, and using this tweak, a recipient is able to calculate the shared secret, because, you know, the shared secret was the tweak and the public key sum and the input hash.

> So this is just enough for any silent payment receiver to create or calculate the actual resulting on chain address.

Reviewer: ```suggestion
So this is just enough for any silent payment receiver to create or calculate the actual resulting on-chain address.
```

It's everyone that uses a wallet, they sort of have to decide if this is worth it for them.
But, yeah, I think these results are pretty nice.
I will say one caveat is that this was all using a blindbit that was running on my same network, so basically this is more of a benchmark that shows the computational side, because we assume that if a blindbit is running on your local network, all of the network stuff is kind of instant.
So this is just to show how long the processing of data looks.
So maybe your results may vary, but they always vary, of course, because you can have different hardware, you can have different network connection speeds.
So it does give a pretty decent picture, at least, I think, of the processing time.

## Future Improvements with Silent Payments

Cygnet: 00:25:10

So yeah, I think, oh, yeah, finally, I just wanted to say, okay, now we have kind of a working light client silent payments wallet.
Because we now have sort of access to this reusable address, we can finally start to think about other things that we can do with, like, a reusable payment code.
So yeah, one of the pretty obvious ones is BIP353.
So BIP353 is like this email address-looking format.

> This is already in use, I think, for example, by Phoenix Wallet, this is pretty popular, but it's also possible, especially with something like silent payment, it's possible now to also do this on chain.

Reviewer: ```suggestion
This is already in use, I think, for example, by Phoenix Wallet, this is pretty popular, but it's also possible, especially with something like silent payment, it's possible now to also do this on-chain.
```
